From 4e68da29ebc45a41845d7127979878930b4c170b Mon Sep 17 00:00:00 2001
From: Awais Belal <awais_belal@mentor.com>
Date: Mon, 5 Sep 2016 15:47:16 +0500
Subject: [PATCH 1/2] demos: make shader location relative

The demo binaries expect the shader (frag/vert.spv)
location to be PWD so a user has to cd to /usr/bin
if the binaries are installed there in order to
run them correctly.
This patch tries to find the location of the binary
and then assumes that the shaders are located in the
same location as the binary so a user can install
everything to a single dir and that will work.

Signed-off-by: Awais Belal <awais_belal@mentor.com>
---
 demos/cube.c | 64 ++++++++++++++++++++++++++++++++++++++++++++++++++++++------
 demos/tri.c  | 49 ++++++++++++++++++++++++++++++++++++++++++++--
 2 files changed, 105 insertions(+), 8 deletions(-)

Index: git/demos/cube.c
===================================================================
--- git.orig/demos/cube.c	2017-02-22 14:31:20.665431016 -0600
+++ git/demos/cube.c	2017-02-22 14:43:02.000000000 -0600
@@ -30,6 +30,7 @@
 #include <stdbool.h>
 #include <assert.h>
 #include <signal.h>
+#include <unistd.h>
 #if defined(VK_USE_PLATFORM_XLIB_KHR) || defined(VK_USE_PLATFORM_XCB_KHR)
 #include <X11/Xutil.h>
 #endif
@@ -410,6 +411,8 @@
 
     uint32_t current_buffer;
     uint32_t queue_family_count;
+
+    char bin_path[255];
 };
 
 VKAPI_ATTR VkBool32 VKAPI_CALL
@@ -1304,18 +1307,25 @@
     const VkFormat tex_format = VK_FORMAT_R8G8B8A8_UNORM;
     VkFormatProperties props;
     uint32_t i;
+    char tex_file[255];
 
     vkGetPhysicalDeviceFormatProperties(demo->gpu, tex_format, &props);
 
     for (i = 0; i < DEMO_TEXTURE_COUNT; i++) {
         VkResult U_ASSERT_ONLY err;
-
+        if (strlen(demo->bin_path) > 0) {
+            strcpy(tex_file, demo->bin_path);
+            strcat(tex_file, "/");
+            strcat(tex_file, tex_files[i]);
+        }
+        else
+            strcpy(tex_file, tex_files[i]);
         if ((props.linearTilingFeatures &
              VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT) &&
             !demo->use_staging_buffer) {
             /* Device can texture using linear textures */
             demo_prepare_texture_image(
-                demo, tex_files[i], &demo->textures[i], VK_IMAGE_TILING_LINEAR,
+                demo, tex_file, &demo->textures[i], VK_IMAGE_TILING_LINEAR,
                 VK_IMAGE_USAGE_SAMPLED_BIT,
                 VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
                     VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
@@ -1332,13 +1342,13 @@
 
             memset(&demo->staging_texture, 0, sizeof(demo->staging_texture));
             demo_prepare_texture_image(
-                demo, tex_files[i], &demo->staging_texture, VK_IMAGE_TILING_LINEAR,
+                demo, tex_files, &demo->staging_texture, VK_IMAGE_TILING_LINEAR,
                 VK_IMAGE_USAGE_TRANSFER_SRC_BIT,
                 VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
                     VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
 
             demo_prepare_texture_image(
-                demo, tex_files[i], &demo->textures[i], VK_IMAGE_TILING_OPTIMAL,
+                demo, tex_files, &demo->textures[i], VK_IMAGE_TILING_OPTIMAL,
                 (VK_IMAGE_USAGE_TRANSFER_DST_BIT | VK_IMAGE_USAGE_SAMPLED_BIT),
                 VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
 
@@ -1676,8 +1686,16 @@
 #else
     void *vertShaderCode;
     size_t size;
+    char fname[255];
+    if (strlen(demo->bin_path) > 0) {
+        strcpy(fname, demo->bin_path);
+        strcat(fname, "/");
+        strcat(fname, "cube-vert.spv");
+    }
+    else
+        strcpy(fname, "cube-vert.spv");
 
-    vertShaderCode = demo_read_spv("cube-vert.spv", &size);
+    vertShaderCode = demo_read_spv(fname, &size);
 
     demo->vert_shader_module =
         demo_prepare_shader_module(demo, vertShaderCode, size);
@@ -1701,8 +1719,16 @@
 #else
     void *fragShaderCode;
     size_t size;
+    char fname[255];
+    if (strlen(demo->bin_path) > 0) {
+        strcpy(fname, demo->bin_path);
+        strcat(fname, "/");
+        strcat(fname, "cube-frag.spv");
+    }
+    else
+        strcpy(fname, "cube-frag.spv");
 
-    fragShaderCode = demo_read_spv("cube-frag.spv", &size);
+    fragShaderCode = demo_read_spv(fname, &size);
 
     demo->frag_shader_module =
         demo_prepare_shader_module(demo, fragShaderCode, size);
@@ -3204,6 +3230,30 @@
 #endif
 }
 
+static void find_bin_path(char *in_arg, char *ret_path) {
+    char *ptr = getenv("PATH");
+    char *pch_temp;
+    char pch[255];
+
+    if (access(in_arg, F_OK ) == 0) {
+        pch_temp = strrchr(in_arg, '/');
+        strncpy(ret_path, in_arg, strlen(in_arg) - strlen(pch_temp));
+    } else if (in_arg[0] != '/') {
+        pch_temp = strtok(ptr, ":");
+        while (pch_temp != NULL) {
+            strcpy(pch, pch_temp);
+            strcat(pch, "/");
+            strcat(pch, in_arg);
+            if ((access(pch, F_OK ) == 0)) {
+                strcpy(ret_path, pch_temp);
+                break;
+            }
+            else
+                pch_temp = strtok(NULL, ":");
+        }
+    }
+}
+
 static void demo_init(struct demo *demo, int argc, char **argv) {
     vec3 eye = {0.0f, 3.0f, 5.0f};
     vec3 origin = {0, 0, 0};
@@ -3280,6 +3330,8 @@
     mat4x4_identity(demo->model_matrix);
 
     demo->projection_matrix[1][1]*=-1;  //Flip projection matrix from GL to Vulkan orientation.
+
+    find_bin_path(argv[0], demo->bin_path);
 }
 
 #if defined(VK_USE_PLATFORM_WIN32_KHR)
